<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>tidyomics: Enhancing Omic Data Analyses - 1&nbsp; Genomic intervals data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/genomic-interaction.html" rel="next">
<link href="../pages/preamble.html" rel="prev">
<link href="../assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/genomic-interval.html">Fundamentals concepts</a></li><li class="breadcrumb-item"><a href="../pages/genomic-interval.html"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Genomic intervals data</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">tidyomics: Enhancing Omic Data Analyses</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/BiocBook.tidyomics/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/BiocBook.tidyomics/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">What is <code>tidyomics</code>?</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/preamble.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Fundamentals concepts</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/genomic-interval.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Genomic intervals data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/genomic-interaction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Genomic interactions data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/summarized-experiment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Summarized experiment data</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Specific omics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/transcriptomic.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Transcriptomic data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/epigenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Epigenomic data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/single-cell.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Single-cell data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/spatial-single-cell.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Spatial single-cell data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/flow-cytometry.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Flow cytometry data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/mass-cytometry.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Mass cytometry data</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Additional resources</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/packages-list.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">List of packages included in the <code>tidyomics</code> framework</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/future-directions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Future directions</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/tidyomics-contributors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">tidyomics contributors</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#importing-granges-from-files" id="toc-importing-granges-from-files" class="nav-link active" data-scroll-target="#importing-granges-from-files"><span class="header-section-number">1.1</span> Importing <code>GRanges</code> from files</a></li>
  <li><a href="#manipulating-granges-with-tidy-verbs" id="toc-manipulating-granges-with-tidy-verbs" class="nav-link" data-scroll-target="#manipulating-granges-with-tidy-verbs"><span class="header-section-number">1.2</span> Manipulating <code>GRanges</code> with tidy verbs</a></li>
  <li><a href="#resources" id="toc-resources" class="nav-link" data-scroll-target="#resources">Resources</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session info</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/genomic-interval.html">Fundamentals concepts</a></li><li class="breadcrumb-item"><a href="../pages/genomic-interval.html"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Genomic intervals data</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Genomic intervals data</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><section id="importing-granges-from-files" class="level2" data-number="1.1"><h2 data-number="1.1" class="anchored" data-anchor-id="importing-granges-from-files">
<span class="header-section-number">1.1</span> Importing <code>GRanges</code> from files</h2>
<p><em>Recommended time: 10 min</em></p>
<p>The <code><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">BiocIO::import()</a></code> generic function lets one import <code>bed</code> files (or alike) into <code>GRanges</code> object in R.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bedf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'S288C-borders.bed'</span>, package <span class="op">=</span> <span class="st">'Bioc2024tidyWorkshop'</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">import</span><span class="op">(</span><span class="va">bedf</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The tidy way of importing genomic ranges in R, however, is to read files as a <code>data.frame</code> (or <code>tibble</code>) and <em>coerce</em> it as a <code>GRanges</code> once it’s imported in memory.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">tib</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span><span class="op">(</span><span class="va">bedf</span>, col_names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tib</span> </span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-construct.html">as_granges</a></span><span class="op">(</span><span class="va">tib</span>, seqnames <span class="op">=</span> <span class="va">X1</span>, start <span class="op">=</span> <span class="va">X2</span>, end <span class="op">=</span> <span class="va">X3</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gr</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note how refering to column names is done using <a href="https://dplyr.tidyverse.org/articles/programming.html#data-masking">tidy evaluation</a>.</p>
</section><section id="manipulating-granges-with-tidy-verbs" class="level2" data-number="1.2"><h2 data-number="1.2" class="anchored" data-anchor-id="manipulating-granges-with-tidy-verbs">
<span class="header-section-number">1.2</span> Manipulating <code>GRanges</code> with tidy verbs</h2>
<p><em>Recommended time: 10 min</em></p>
<p>With <code>plyranges</code>, <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1597-8/tables/1">a number of tidy operations</a> are readily available, e.g.:</p>
<ul>
<li><code><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code></li>
<li><code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code></li>
<li><code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code></li>
<li>
<code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> + <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize()</a></code>
</li>
<li>…</li>
</ul>
<p>Just like with <code>tidyverse</code> functions, these operations work well with the native <code>|&gt;</code> pipe in R.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gr</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">score</span> <span class="op">&gt;</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>round_score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">score</span>, digits <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">round_score</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html">mean</a></span><span class="op">(</span><span class="va">score</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>But really, what <code>plyranges</code> excels at is providing direct access to “protected” variables, i.e.&nbsp;<code>seqnames</code>, <code>start</code>, <code>end</code>, <code>width</code>, <code>strand</code>, …</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gr</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>        seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="st">'XVI'</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html">levels</a></span><span class="op">(</span><span class="va">seqnames</span><span class="op">)</span><span class="op">)</span>, </span>
<span>        width <span class="op">=</span> <span class="fl">1</span>, </span>
<span>        strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'-'</span>, <span class="st">'+'</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, a number of operations for genomic arithmetics are provided by <code>plyranges</code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gr</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-anchor.html">anchor_center</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/stretch.html">stretch</a></span><span class="op">(</span>extend <span class="op">=</span> <span class="op">-</span><span class="fl">1000</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-shift.html">shift_upstream</a></span><span class="op">(</span><span class="fl">250</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-flank.html">flank_upstream</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="resources" class="level1 unnumbered"><h1 class="unnumbered">Resources</h1>
<ul>
<li><a href="https://tidyomics.github.io/tidy-ranges-tutorial/">“Tidy Ranges Tutorial” by Michael Love</a></li>
<li><a href="https://jserizay.com/Bioc2024tidyworkshop/">A Bioc2024 workshop on plyranges and others</a></li>
</ul></section><section id="session-info" class="level1 unnumbered"><h1 class="unnumbered">Session info</h1>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to expand 👇
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">sessioninfo</span><span class="fu">::</span><span class="fu"><a href="https://r-lib.github.io/sessioninfo/reference/session_info.html">session_info</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span>,<span class="st">"Package"</span><span class="op">]</span>, </span>
<span>    include_base <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">##  ─ Session info ────────────────────────────────────────────────────────────</span></span>
<span><span class="co">##   setting  value</span></span>
<span><span class="co">##   version  R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">##   os       Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">##   system   x86_64, linux-gnu</span></span>
<span><span class="co">##   ui       X11</span></span>
<span><span class="co">##   language (EN)</span></span>
<span><span class="co">##   collate  C</span></span>
<span><span class="co">##   ctype    en_US.UTF-8</span></span>
<span><span class="co">##   tz       Etc/UTC</span></span>
<span><span class="co">##   date     2024-07-27</span></span>
<span><span class="co">##   pandoc   3.2 @ /usr/bin/ (via rmarkdown)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  ─ Packages ────────────────────────────────────────────────────────────────</span></span>
<span><span class="co">##   package                  * version    date (UTC) lib source</span></span>
<span><span class="co">##   abind                      1.4-5      2016-07-21 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   askpass                    1.2.0      2023-09-03 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   backports                  1.5.0      2024-05-23 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   base                     * 4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   base64enc                  0.1-3      2015-07-28 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   BH                         1.84.0-0   2024-01-10 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   Biobase                    2.65.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   BiocBook.tidyomics         0.98.1     2024-07-27 [1] local</span></span>
<span><span class="co">##   BiocFileCache              2.13.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   BiocGenerics               0.51.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   BiocIO                     1.15.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   BiocManager                1.30.23    2024-05-04 [2] CRAN (R 4.4.1)</span></span>
<span><span class="co">##   BiocParallel               1.39.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   BiocStyle                  2.33.1     2024-06-12 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   BiocVersion                3.20.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.1)</span></span>
<span><span class="co">##   Biostrings                 2.73.1     2024-06-02 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   bit                        4.0.5      2022-11-15 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   bit64                      4.0.5      2020-08-30 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   bitops                     1.0-7      2021-04-24 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   blob                       1.2.4      2023-03-17 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   bookdown                   0.40       2024-07-02 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   boot                       1.3-30     2024-02-26 [3] CRAN (R 4.4.1)</span></span>
<span><span class="co">##   brew                       1.0-10     2023-12-16 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   brio                       1.1.5      2024-04-24 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   broom                      1.0.6      2024-05-17 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   bslib                      0.7.0      2024-03-29 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   cachem                     1.1.0      2024-05-16 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   callr                      3.7.6      2024-03-25 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   caTools                    1.18.2     2021-03-28 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   cellranger                 1.1.0      2016-07-27 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   class                      7.3-22     2023-05-03 [3] CRAN (R 4.4.1)</span></span>
<span><span class="co">##   cli                        3.6.3      2024-06-21 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   clipr                      0.8.0      2022-02-22 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   clock                      0.7.1      2024-07-18 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   cluster                    2.1.6      2023-12-01 [3] CRAN (R 4.4.1)</span></span>
<span><span class="co">##   codetools                  0.2-20     2024-03-31 [3] CRAN (R 4.4.1)</span></span>
<span><span class="co">##   colorspace                 2.1-0      2023-01-23 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   commonmark                 1.9.1      2024-01-30 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   compiler                   4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   conflicted                 1.2.0      2023-02-01 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   cowplot                    1.1.3      2024-01-22 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   cpp11                      0.4.7      2023-12-02 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   crayon                     1.5.3      2024-06-20 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   credentials                2.0.1      2023-09-06 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   crosstalk                  1.2.1      2023-11-23 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   curl                       5.2.1      2024-03-01 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   cytolib                    2.17.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   data.table                 1.15.4     2024-03-30 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   datasets                 * 4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   DBI                        1.2.3      2024-06-02 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   dbplyr                     2.5.0      2024-03-19 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   DelayedArray               0.31.9     2024-07-17 [2] Bioconductor 3.20 (R 4.4.1)</span></span>
<span><span class="co">##   deldir                     2.0-4      2024-02-28 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   desc                       1.4.3      2023-12-10 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   devtools                   2.4.5      2022-10-11 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   diagram                    1.6.5      2020-09-30 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   diffobj                    0.3.5      2021-10-05 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   digest                     0.6.36     2024-06-23 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   docopt                     0.7.1      2020-06-24 [2] RSPM (R 4.4.1)</span></span>
<span><span class="co">##   doParallel                 1.0.17     2022-02-07 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   dotCall64                  1.1-1      2023-11-28 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   downlit                    0.4.4      2024-06-10 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   dplyr                      1.1.4      2023-11-17 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   dqrng                      0.4.1      2024-05-28 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   dtplyr                     1.3.1      2023-03-22 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   easylift                   1.3.0      2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   ellipsis                   0.3.2      2021-04-29 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   evaluate                   0.24.0     2024-06-10 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   fansi                      1.0.6      2023-12-08 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   farver                     2.1.2      2024-05-13 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   fastDummies                1.7.3      2023-07-06 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   fastmap                    1.2.0      2024-05-15 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   filelock                   1.0.3      2023-12-11 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   fitdistrplus               1.2-1      2024-07-12 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   flowCore                   2.17.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   FNN                        1.1.4      2024-01-12 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   fontawesome                0.5.2      2023-08-19 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   forcats                    1.0.0      2023-01-29 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   foreach                    1.5.2      2022-02-02 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   foreign                    0.8-87     2024-06-26 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   formatR                    1.14       2023-01-17 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   fs                         1.6.4      2024-04-25 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   futile.logger              1.4.3      2016-07-10 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   futile.options             1.0.1      2018-04-20 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   future                     1.33.2     2024-03-26 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   future.apply               1.11.2     2024-03-28 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   gargle                     1.5.2      2023-07-20 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   generics                   0.1.3      2022-07-05 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   GenomeInfoDb               1.41.1     2024-05-24 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   GenomeInfoDbData           1.2.12     2024-06-24 [2] Bioconductor</span></span>
<span><span class="co">##   GenomicAlignments          1.41.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   GenomicRanges              1.57.1     2024-06-12 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   gert                       2.1.0      2024-07-19 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   ggforce                    0.4.2      2024-02-19 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   ggplot2                    3.5.1      2024-04-23 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   ggraph                     2.2.1      2024-03-07 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   ggrepel                    0.9.5      2024-01-10 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   ggridges                   0.5.6      2024-01-23 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   gh                         1.4.1      2024-03-28 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   gitcreds                   0.1.2      2022-09-08 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   glmnet                     4.1-8      2023-08-22 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   globals                    0.16.3     2024-03-08 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   glue                       1.7.0      2024-01-09 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   goftest                    1.2-3      2021-10-07 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   googledrive                2.1.1      2023-06-11 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   googlesheets4              1.1.1      2023-06-11 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   gower                      1.0.1      2022-12-22 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   gplots                     3.1.3.1    2024-02-02 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   graphics                 * 4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   graphlayouts               1.1.1      2024-03-09 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   grDevices                * 4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   grid                       4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   gridExtra                  2.3        2017-09-09 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   gtable                     0.3.5      2024-04-22 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   gtools                     3.9.5      2023-11-20 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   hardhat                    1.4.0      2024-06-02 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   haven                      2.5.4      2023-11-30 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   here                       1.0.1      2020-12-13 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   highr                      0.11       2024-05-26 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   hms                        1.1.3      2023-03-21 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   htmltools                  0.5.8.1    2024-04-04 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   htmlwidgets                1.6.4      2023-12-06 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   httpuv                     1.6.15     2024-03-26 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   httr                       1.4.7      2023-08-15 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   httr2                      1.0.2      2024-07-16 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   ica                        1.0-3      2022-07-08 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   ids                        1.0.1      2017-05-31 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   igraph                     2.0.3      2024-03-13 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   ini                        0.3.1      2018-05-20 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   InteractionSet             1.33.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   ipred                      0.9-15     2024-07-18 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   IRanges                    2.39.2     2024-07-17 [2] Bioconductor 3.20 (R 4.4.1)</span></span>
<span><span class="co">##   irlba                      2.3.5.1    2022-10-03 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   isoband                    0.2.7      2022-12-20 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   iterators                  1.0.14     2022-02-05 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   jquerylib                  0.1.4      2021-04-26 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   jsonlite                   1.8.8      2023-12-04 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   KernSmooth                 2.23-24    2024-05-17 [3] CRAN (R 4.4.1)</span></span>
<span><span class="co">##   knitr                      1.48       2024-07-07 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   labeling                   0.4.3      2023-08-29 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   lambda.r                   1.2.4      2019-09-18 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   later                      1.3.2      2023-12-06 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   lattice                    0.22-6     2024-03-20 [3] CRAN (R 4.4.1)</span></span>
<span><span class="co">##   lava                       1.8.0      2024-03-05 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   lazyeval                   0.2.2      2019-03-15 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   leiden                     0.4.3.1    2023-11-17 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   lifecycle                  1.0.4      2023-11-07 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   listenv                    0.9.1      2024-01-29 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   littler                    0.3.20     2024-03-23 [2] RSPM (R 4.4.1)</span></span>
<span><span class="co">##   lmtest                     0.9-40     2022-03-21 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   lubridate                  1.9.3      2023-09-27 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   magick                     2.8.4      2024-07-14 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   magrittr                   2.0.3      2022-03-30 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   MASS                       7.3-61     2024-06-13 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   Matrix                     1.7-0      2024-04-26 [3] CRAN (R 4.4.1)</span></span>
<span><span class="co">##   MatrixGenerics             1.17.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   matrixStats                1.3.0      2024-04-11 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   memoise                    2.0.1      2021-11-26 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   methods                  * 4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   mgcv                       1.9-1      2023-12-21 [3] CRAN (R 4.4.1)</span></span>
<span><span class="co">##   mime                       0.12       2021-09-28 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   miniUI                     0.1.1.1    2018-05-18 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   modelr                     0.1.11     2023-03-22 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   munsell                    0.5.1      2024-04-01 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   nlme                       3.1-165    2024-06-06 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   nnet                       7.3-19     2023-05-03 [3] CRAN (R 4.4.1)</span></span>
<span><span class="co">##   nullranges                 1.11.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   numDeriv                   2016.8-1.1 2019-06-06 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   openssl                    2.2.0      2024-05-16 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   parallel                   4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   parallelly                 1.37.1     2024-02-29 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   patchwork                  1.2.0      2024-01-08 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   pbapply                    1.7-2      2023-06-27 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   pillar                     1.9.0      2023-03-22 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   pkgbuild                   1.4.4      2024-03-17 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   pkgconfig                  2.0.3      2019-09-22 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   pkgdown                    2.1.0      2024-07-06 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   pkgload                    1.4.0      2024-06-28 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   plogr                      0.2.0      2018-03-25 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   plotly                     4.10.4     2024-01-13 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   plyinteractions            1.3.1      2024-07-10 [2] Bioconductor 3.20 (R 4.4.1)</span></span>
<span><span class="co">##   plyr                       1.8.9      2023-10-02 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   plyranges                  1.25.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   png                        0.1-8      2022-11-29 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   polyclip                   1.10-7     2024-07-23 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   praise                     1.0.0      2015-08-11 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   preprocessCore             1.67.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.1)</span></span>
<span><span class="co">##   prettyunits                1.2.0      2023-09-24 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   processx                   3.8.4      2024-03-16 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   prodlim                    2024.06.25 2024-06-24 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   profvis                    0.3.8      2023-05-02 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   progress                   1.2.3      2023-12-06 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   progressr                  0.14.0     2023-08-10 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   promises                   1.3.0      2024-04-05 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   ps                         1.7.7      2024-07-02 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   purrr                      1.0.2      2023-08-10 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   R.methodsS3                1.8.2      2022-06-13 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   R.oo                       1.26.0     2024-01-24 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   R.utils                    2.12.3     2023-11-18 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   R6                         2.5.1      2021-08-19 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   ragg                       1.3.2      2024-05-15 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   RANN                       2.6.1      2019-01-08 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   rappdirs                   0.3.3      2021-01-31 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   rcmdcheck                  1.4.0      2021-09-27 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   RColorBrewer               1.1-3      2022-04-03 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   Rcpp                       1.0.13     2024-07-17 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   RcppAnnoy                  0.0.22     2024-01-23 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   RcppArmadillo              14.0.0-1   2024-07-08 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   RcppEigen                  0.3.4.0.0  2024-02-28 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   RcppHNSW                   0.6.0      2024-02-04 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   RcppProgress               0.4.2      2020-02-06 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   RcppTOML                   0.2.2      2023-01-29 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   RCurl                      1.98-1.16  2024-07-11 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   readr                      2.1.5      2024-01-10 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   readxl                     1.4.3      2023-07-06 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   recipes                    1.1.0      2024-07-04 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   rematch                    2.0.0      2023-08-30 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   rematch2                   2.1.2      2020-05-01 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   remotes                    2.5.0      2024-03-17 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   reprex                     2.1.1      2024-07-06 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   reshape2                   1.4.4      2020-04-09 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   restfulr                   0.0.15     2022-06-16 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   reticulate                 1.38.0     2024-06-19 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   Rhdf5lib                   1.27.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   Rhtslib                    3.1.0      2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   rjson                      0.2.21     2022-01-09 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   rlang                      1.1.4      2024-06-04 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   rmarkdown                  2.27       2024-05-17 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   ROCR                       1.0-11     2020-05-02 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   roxygen2                   7.3.2      2024-06-28 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   rpart                      4.1.23     2023-12-05 [3] CRAN (R 4.4.1)</span></span>
<span><span class="co">##   rprojroot                  2.0.4      2023-11-05 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   RProtoBufLib               2.17.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   Rsamtools                  2.21.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   RSpectra                   0.16-2     2024-07-18 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   RSQLite                    2.3.7      2024-05-27 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   rstudioapi                 0.16.0     2024-03-24 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   rtracklayer                1.65.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   Rtsne                      0.17       2023-12-07 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   rversions                  2.1.2      2022-08-31 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   rvest                      1.0.4      2024-02-12 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   S4Arrays                   1.5.5      2024-07-21 [2] Bioconductor 3.20 (R 4.4.1)</span></span>
<span><span class="co">##   S4Vectors                  0.43.2     2024-07-17 [2] Bioconductor 3.20 (R 4.4.1)</span></span>
<span><span class="co">##   sass                       0.4.9      2024-03-15 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   scales                     1.3.0      2023-11-28 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   scattermore                1.2        2023-06-12 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   sctransform                0.4.1      2023-10-19 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   selectr                    0.4-2      2019-11-20 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   sessioninfo                1.2.2      2021-12-06 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   Seurat                     5.1.0      2024-05-10 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   SeuratObject               5.0.2      2024-05-08 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   shape                      1.4.6.1    2024-02-23 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   shiny                      1.8.1.1    2024-04-02 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   SingleCellExperiment       1.27.2     2024-05-24 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   sitmo                      2.0.2      2021-10-13 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   snow                       0.4-4      2021-10-27 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   sourcetools                0.1.7-1    2023-02-01 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   sp                         2.1-4      2024-04-30 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   spam                       2.10-0     2023-10-23 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   SparseArray                1.5.26     2024-07-25 [2] Bioconductor 3.20 (R 4.4.1)</span></span>
<span><span class="co">##   spatial                    7.3-17     2023-07-20 [3] CRAN (R 4.4.1)</span></span>
<span><span class="co">##   SpatialExperiment          1.15.1     2024-06-20 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   spatstat.data              3.1-2      2024-06-21 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   spatstat.explore           3.3-1      2024-07-15 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   spatstat.geom              3.3-2      2024-07-15 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   spatstat.random            3.3-1      2024-07-15 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   spatstat.sparse            3.1-0      2024-06-21 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   spatstat.univar            3.0-0      2024-06-28 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   spatstat.utils             3.0-5      2024-06-17 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   splines                    4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   SQUAREM                    2021.1     2021-01-13 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   stats                    * 4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   stats4                     4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   stringi                    1.8.4      2024-05-06 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   stringr                    1.5.1      2023-11-14 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   SummarizedExperiment       1.35.1     2024-06-28 [2] Bioconductor 3.20 (R 4.4.1)</span></span>
<span><span class="co">##   survival                   3.7-0      2024-06-05 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   sys                        3.4.2      2023-05-23 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   systemfonts                1.1.0      2024-05-15 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   tcltk                      4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   tensor                     1.5        2012-05-05 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   testthat                   3.2.1.1    2024-04-14 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   textshaping                0.4.0      2024-05-24 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   tibble                     3.2.1      2023-03-20 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   tidybulk                   1.17.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   tidyFlowCore               0.99.2     2024-05-17 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   tidygate                   1.0.13     2024-07-27 [2] Github (stemangiola/tidygate@2319f9f)</span></span>
<span><span class="co">##   tidygraph                  1.3.1      2024-01-30 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   tidyomics                  1.1.0      2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   tidyr                      1.3.1      2024-01-24 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   tidyselect                 1.2.1      2024-03-11 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   tidyseurat                 0.8.0      2024-01-10 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   tidySingleCellExperiment   1.15.1     2024-05-15 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   tidySpatialExperiment      1.1.4      2024-07-14 [2] Bioconductor 3.20 (R 4.4.1)</span></span>
<span><span class="co">##   tidySummarizedExperiment   1.15.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   tidytof                    0.99.7     2024-07-27 [2] Github (keyes-timothy/tidytof@f94e11a)</span></span>
<span><span class="co">##   tidyverse                  2.0.0      2023-02-22 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   timechange                 0.3.0      2024-01-18 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   timeDate                   4032.109   2023-12-14 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   tinytex                    0.52       2024-07-18 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   tools                      4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   ttservice                  0.4.1      2024-06-07 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   tweenr                     2.0.3      2024-02-26 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   tzdb                       0.4.0      2023-05-12 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   UCSC.utils                 1.1.0      2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   urlchecker                 1.0.1      2021-11-30 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   usethis                    2.2.3      2024-02-19 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   utf8                       1.2.4      2023-10-22 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   utils                    * 4.4.1      2024-06-14 [3] local</span></span>
<span><span class="co">##   uuid                       1.2-0      2024-01-14 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   uwot                       0.2.2      2024-04-21 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   vctrs                      0.6.5      2023-12-01 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   viridis                    0.6.5      2024-01-29 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   viridisLite                0.4.2      2023-05-02 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   vroom                      1.6.5      2023-12-05 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   waldo                      0.5.2      2023-11-02 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   whisker                    0.4.1      2022-12-05 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   withr                      3.0.0      2024-01-16 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   xfun                       0.46       2024-07-18 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   XML                        3.99-0.17  2024-06-25 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   xml2                       1.3.6      2023-12-04 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   xopen                      1.0.1      2024-04-25 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   xtable                     1.8-4      2019-04-21 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   XVector                    0.45.0     2024-05-01 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   yaml                       2.3.9      2024-07-05 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   yardstick                  1.3.1      2024-03-21 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   zip                        2.3.1      2024-01-27 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##   zlibbioc                   1.51.1     2024-06-05 [2] Bioconductor 3.20 (R 4.4.0)</span></span>
<span><span class="co">##   zoo                        1.8-12     2023-04-13 [2] RSPM (R 4.4.0)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##   [1] /tmp/RtmpIc2SKq/Rinst8c18aa7b11</span></span>
<span><span class="co">##   [2] /usr/local/lib/R/site-library</span></span>
<span><span class="co">##   [3] /usr/local/lib/R/library</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  ─ Python configuration ────────────────────────────────────────────────────</span></span>
<span><span class="co">##   Python is not available</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  ───────────────────────────────────────────────────────────────────────────</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section><section id="references" class="level1 unnumbered"><h1 class="unnumbered">References</h1>


</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/preamble.html" class="pagination-link" aria-label="Preamble">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Preamble</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/genomic-interaction.html" class="pagination-link" aria-label="Genomic interactions data">
        <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Genomic interactions data</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


</body></html>